### PR TITLE
feat: add TDM（time division multiplexing) plugin

### PR DESCRIPTION
**Pull request by yahaa**

### CODE REVIEW

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+	"volcano.sh/volcano/pkg/scheduler/api"
+	"volcano.sh/volcano/pkg/scheduler/framework"
+)
+
+const (
+	// PluginName indicates name of volcano scheduler plugin.
+	PluginName = "tdm"
+	// PreemptableTaskKey pod annotation key
+	PreemptableTaskKey = "volcano.sh/preemptable"
```
Comment (k82cn) [inline pkg/scheduler/plugins/tdm/tdm.go:33]: prefer to move them to global const, so other plugins may also reuse this one :)

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		if !ok {
+			return score, nil
+		}
+
+		if err := bp.availableRevocableZone(rz, ssn); err != nil {
+			klog.V(4).Infof("TDM plugin rz not available %s", err)
+			return score, nil
+		}
+
+		if _, ok := task.Pod.Annotations[PreemptableTaskKey]; !ok {
```
Comment (k82cn) [inline pkg/scheduler/plugins/tdm/tdm.go:77]: Let's add `Preemptable` to `TaskInfo` :)

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+func (bp *tdmPlugin) OnSessionOpen(ssn *framework.Session) {
+	klog.V(4).Infof("Enter tdm plugin ...")
+	if klog.V(4) {
+		defer func() {
+			klog.V(4).Infof("Leaving tdm plugin.")
+		}()
+	}
+
+	// tdm plugin just handle revocable node
+	nodeOrderFn := func(task *api.TaskInfo, node *api.NodeInfo) (float64, error) {
```
Comment (k82cn) [inline pkg/scheduler/plugins/tdm/tdm.go:64]: hm... in your case, `nodeOrderFn` maybe unnecessary :)

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+			victims = append(victims, item)
+		}
+
+		klog.V(4).Infof("TDM victims are %+v", victims)
+
+		return victims
+	}
+
+	victimsFn := func() []*api.TaskInfo {
```
Comment (wpeng102) [inline pkg/scheduler/plugins/tdm/tdm.go:190]: we need support expel step by step, so in victimsFn not return all the victim `taskInfo` in one scheduler period.

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 1 above)
Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go]: Moved it into `pkg/scheduler/api/well-known-labels. go`

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 2 above)
Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go]: Ok, it done.

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 3 above)
Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go:175]: I think it was necessary, preemptable task can schedule to all kind of node, revocable node just have the high priority.

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 4 above)
Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go:229]: You are right.

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 3 above)
Comment (k82cn) [reply pkg/scheduler/plugins/tdm/tdm.go:175]: ok, great!

File: pkg/scheduler/api/pod_info.go [hunk 5]
```diff
@@ -67,6 +68,18 @@ func GetPodResourceRequest(pod *v1.Pod) *Resource {
 	return result
 }
 
+// GetPodPreemptable return volcano.sh/preemptable value for pod
+func GetPodPreemptable(pod *v1.Pod) bool {
+	if len(pod.Annotations) > 0 {
+		if value, found := pod.Annotations[v1beta1.PodPreemptable]; found {
+			if value == "true" {
```
Comment (k82cn) [inline pkg/scheduler/api/pod_info.go:75]: prefer to use `strconv.ParseBool`

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+			if err := bp.availableRevocableZone(rz); err != nil {
+				// rz disactive, then evict preemptable tasks from the revocable node
+				victims = append(victims, bp.revocableNodePreemptableTask(rz, ssn)...)
+			} else {
+				active[rz] = struct{}{}
+			}
+		}
+
+		// if there are some active rz, then evict preemptable tasks from all the none revocable node
+		victims = append(victims, bp.noneRevocableNodePreemptableTask(ssn)...)
```
Comment (wpeng102) [inline pkg/scheduler/plugins/tdm/tdm.go:250]: It seems there is no check if has active rz. If there is no active rz, the preemtable pod on non revocable zone will be evicted again and again.

Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go]: Yes, fixed.

**Comment by wpeng102** [comment]
suggest use the following configure to test:
```
actions: "enqueue, allocate, backfill, preempt"
    tiers:
    - plugins:
      - name: tdm
        arguments:
          tdm.revocableZone.rz1: 1:00-4:00
          tdm.evict.period: 1m
    - plugins:
      - name: priority
      - name: gang
      - name: conformance
    - plugins:
      - name: drf
      - name: predicates
      - name: proportion
      - name: nodeorder
      - name: binpack

```

**Comment by yahaa** [comment]
> suggest use the following configure to test:
> 
> ```
> actions: "enqueue, allocate, backfill, preempt"
>     tiers:
>     - plugins:
>       - name: tdm
>         arguments:
>           tdm.revocableZone.rz1: 1:00-4:00
>           tdm.evict.period: 1m
>     - plugins:
>       - name: priority
>       - name: gang
>       - name: conformance
>     - plugins:
>       - name: drf
>       - name: predicates
>       - name: proportion
>       - name: nodeorder
>       - name: binpack
> ```

Got it.

**Comment by wpeng102** [comment]
The framework and basic funtion is OK to me. Suggest merge this PR first.

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+	ssn.AddPredicateFn(bp.Name(), predicateFn)
+	ssn.AddNodeOrderFn(bp.Name(), nodeOrderFn)
+	ssn.AddPreemptableFn(bp.Name(), preemptableFn)
+	ssn.AddVictimTasksFns(bp.Name(), victimsFn)
+}
+
+func (bp *tdmPlugin) maxVictims(victims []*api.TaskInfo) []*api.TaskInfo {
+	targetNum := 0
+	switch bp.evictMaxNum.Type {
+	case intstr.Int:
```
Comment (k82cn) [inline pkg/scheduler/plugins/tdm/tdm.go:281]: why check its type? let's convert it to float when got this parameter.

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+		targetNum = int(math.Min(float64(bp.evictMaxNum.IntValue()), float64(len(victims))))
```
Comment (k82cn) [inline pkg/scheduler/plugins/tdm/tdm.go:282]: are we going to select Task/Pod by Job/PodGroup, e.g. not only get pod from one PodGroup.

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+	}
+
+	return victims[:targetNum]
+}
+
+func (bp *tdmPlugin) revocableNodePreemptableTask(rz string, ssn *framework.Session) []*api.TaskInfo {
+	tasks := make([]*api.TaskInfo, 0)
+
+	for _, node := range ssn.Nodes {
+		if node.RevocableZone != rz {
```
Comment (k82cn) [inline pkg/scheduler/plugins/tdm/tdm.go:298]: lets' build an index in cache for revocable nodes.

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+		score = float64(v1alpha1.MaxNodeScore)
+
+		klog.V(4).Infof("TDM score for Task %s/%s on node %s is: %v", task.Namespace, task.Name, node.Name, score)
+		return score, nil
+	}
+
+	preemptableFn := func(preemptor *api.TaskInfo, preemptees []*api.TaskInfo) []*api.TaskInfo {
+		if preemptor.Preemptable {
+			klog.V(4).Infof("TDM task %s/%s is preemptable, do nothing skip", preemptor.Namespace, preemptor.Name)
+			return preemptees
```
Comment (k82cn) [inline pkg/scheduler/plugins/tdm/tdm.go:203]: return nil?

File: pkg/scheduler/plugins/tdm/tdm.go [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+				active[rz] = struct{}{}
+			}
+		}
+
+		// if there are some active rz, then evict preemptable tasks from all the none revocable node
+		if len(active) > 0 {
+			victims = append(victims, bp.noneRevocableNodePreemptableTask(ssn)...)
+		}
+
+		victims = bp.maxVictims(victims)
```
Comment (k82cn) [inline pkg/scheduler/plugins/tdm/tdm.go:239]: It requires to preempt pods in one node, so `maxVictims` is not suitable for this.

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 7 above)
Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go:281]: I want evictMaxNum support two kind of parameter, such as
* evictMaxNum: 5    （this mean that user can specify exactly number of tasks)
* evictMaxNum: 5%  

Dose it make sense?

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 10 above)
Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go:203]: Yes, it should return nil.

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 9 above)
Comment (wpeng102) [reply pkg/scheduler/plugins/tdm/tdm.go:298]: OK，I will use a new PR to handle this comments.

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 11 above)
Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go:239]: Your mean in this preemptableFn we no need to call  `bp.maxVictims()`?

Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go:239]: Yes, it really don't need.

File: pkg/scheduler/plugins/tdm/tdm.go (hunk 8 above)
Comment (yahaa) [reply pkg/scheduler/plugins/tdm/tdm.go:284]: OK，I will use a new PR to handle this comments, after we support evictMaxNum params in workload.

**Pull request merged**